src/TranslatableTrait.php

```diff
@@ -0,0 +1,121 @@
+    }
+
+    /**
+     * Translates the given message.
+     *
+     * When a number is provided as a parameter named "%count%", the message is parsed for plural
+     * forms and a translation is chosen according to this number using the following rules:
+     *
+     * Given a message with different plural translations separated by a
+     * pipe (|), this method returns the correct portion of the message based
+     * on the given number, locale and the pluralization rules in the message
+     * itself.
+     *
+     * The message supports two different types of pluralization rules:
+     *
+     * interval: {0} There are no apples|{1} There is one apple|]1,Inf] There are %count% apples
+     * indexed:  There is one apple|There are %count% apples
+     *
+     * The indexed solution can also contain labels (e.g. one: There is one apple).
+     * This is purely for making the translations more clear - it does not
+     * affect the functionality.
+     *
+     * The two methods can also be mixed:
+     *     {0} There are no apples|one: There is one apple|more: There are %count% apples
+     *
+     * An interval can represent a finite set of numbers:
+     *  {1,2,3,4}
+     *
+     * An interval can represent numbers between two numbers:
+     *  [1, +Inf]
+     *  ]-1,2[
+     *
+     * The left delimiter can be [ (inclusive) or ] (exclusive).
+     * The right delimiter can be [ (exclusive) or ] (inclusive).
+     * Beside numbers, you can use -Inf and +Inf for the infinite.
+     *
+     * @see https://en.wikipedia.org/wiki/ISO_31-11
+     *
+     * @param string      $id         The message id (may also be an object that can be cast to string)
+     * @param array       $parameters An array of parameters for the message
+     * @param string|null $domain     The domain for the message or null to use the default
+     * @param string|null $locale     The locale or null to use the default
+     *
+     * @throws \InvalidArgumentException If the locale contains invalid characters
+     *
+     * @return string The translated string
+     */
+    public function _($id, array $parameters = [], $domain = null, $locale = null) :string
+    {
+        // case if is child Object which use AppScopeTrait + TranslatableTrait
+        // that was already add to an Object which use AppScopeTrait + ContainerTrait + TranslatableTrait
+        if ($this->hasTranslatorInAppScope()) {
+            return $this->app->_($id, $parameters, $domain, $locale);
+        }
+
+        // case if use TranslatableTrait and translator is defined
+        if ($this->hasTranslator()) {
+            return $this->translator->trans($id, $parameters, $domain, $locale);
+        }
+
+        // FALLBACK : Replicate internal library functionality
+        // if not use TranslatableTrait and is not a Child of App which use TranslatableTrait
+        $translator = new class() implements TranslatorInterface {
+            use TranslatorTrait;
+        };
```

Comment:
Kind of wasteful to create a object every time, maybe add a static member, or a private one and initialize that if it's null?

Reply:
Yeah, I think it's a misuse of "trait", in this case - just use of regular class would suffice, although I don't want ATK to handle the case when $app is undefined and should simply `return $id`;

Reply:
I was thinking about if we want to have a fallback class here, but I still believe it's bad, because some users may end up relying on it and it would re-read language files on every invocation, which is an INSANE performance impact.

Either we use $app->translator or rely on service locator pr #99 inside App class. Either way - trait should assume that translator would be pre-initialized with the language files loaded.

Reply:
Well, the fallback class should not ever translate anything, only parse the incoming string, I believe.
As for this specific line of code I'd simply suggest something like
```php
// where self::$_translationFallback = static private member defined to null

if (self::$_translationFallback === null) {
    self::$_translationFallback = new class() implements TranslatorInterface { 
        use TranslatorTrait;
    };
}

// use self::$_translationFallback
```

But that's only assuming that the TranslatorInterface/TranslatorTrait is required to parse the text, not actually translate it, or create this object somehow that it doesn't actually try to load language files.

Reply:
>But that's only assuming that the TranslatorInterface/TranslatorTrait is required to parse the text, not actually translate it, or create this object somehow that it doesn't actually try to load language files.

As you said, it only parse the arguments to return an excepted results.

My idea is that in this way we don't have to delegate to $this->app in case of external use, i think that $_translatorFallback will be called only if the Class not use AppScope or $translator is not set, just to have a fallback for usage of syntax excepted by Symfony\Components\Translation, that's it.

The idea was to use on every Object the same Trait and always use the method _ in place of $this->app, because in this case we have to move everything in AppScope or make 2 trait TranslatableTrait and TranslatorTrait.

If you make it today at 18:00 London, we have a meeting.
I don't know in which timezone you are, but you can come and partecipate, at this link : https://meet.google.com/dfb-vquj-kkk

Reply:
Well yes, I understand why you did it and I prefer having the `_` function available in the classes themselves instead of calling `$this->app` constantly, that does seem like the better way to do it, since you can move any error checking to `_` as well.

I should be able to make it, I'm in +2 UT so it's not really an issue.